src/vanilla.ts

```diff
@@ -241,21 +241,21 @@ export const subscribe = (
   }
 }
 
-export type DeepResolveType<T> = T extends (...args: any[]) => any
+export type Snapshot<T> = T extends (...args: any[]) => any
   ? T
   : T extends AsRef
-  ? T
+  ? Readonly<T>
   : T extends Promise<infer V>
-  ? V
+  ? Readonly<V>
```

Comment:
Actually, there was a bug previously. This has to be recursive.
```suggestion
  ? DeepResolveType<V>
```